feat(ranking): Add RankingModes constants for GetRanking.rankingMode

ashquarky opened the pull request:
<!--

* Before making a pull request, ensure the changes are for an approved issue.
* If your changes are not for an approved issue, your pull request can and will be rejected.
*
* CHECK https://github.com/PretendoNetwork/REPO_NAME/issues?q=is%3Aopen+is%3Aissue+label%3Aapproved
* FOR APPROVED ISSUES!

-->

Part of https://github.com/PretendoNetwork/nex-protocols-common-go/pull/34

### Changes:

Add RankingModes enum for the different GetRanking request modes. This allows downstreams like nex-protocols-common-go to implement things like Friends and Nearby leaderboards.

- [x] I have read and agreed to the [Code of Conduct](https://github.com/PretendoNetwork/Pretendo/blob/master/.github/CODE_OF_CONDUCT.md).
- [x] I have read and complied with the [contributing guidelines](https://github.com/PretendoNetwork/Pretendo/blob/master/.github/CONTRIBUTING.md).
- [ ] What I'm implementing was an [approved issue](../issues?q=is%3Aopen+is%3Aissue+label%3Aapproved).
- [x] I have tested all of my changes.

DaniElectra commented on ranking/get_ranking.go:
```diff
@@ -92,3 +92,25 @@ func (protocol *Protocol) handleGetRanking(packet nex.PacketInterface) {
 	globals.Respond(packet, rmcMessage)
 }
+
+type rankingModes struct {
+	// Global leaderboards
+	Global uint8
+	// Ranks close to caller on global leaderboards
+	NearbyGlobal uint8
+	// Friends of caller only
+	Friends uint8
+	// Ranks close to caller on friends leaderboards
+	NearbyFriends uint8
+	// Rank of caller only
+	Self uint8
+}
+
+// RankingModes is an enum of all the types of ranking request that can be made in the GetRanking method
+var RankingModes = rankingModes{
+	Global:        0,
+	NearbyGlobal:  1,
+	Friends:       2,
+	NearbyFriends: 3,
+	Self:          4,
+}
```
Not sure if this should be on the same file as the `GetRanking` handling? Maybe on a different file named `ranking_modes.go`?

jonbarrow replied:
I agree with Dani. Having it in it's own file would make it line up with all of our other constants definition files, like those in nex-go and https://github.com/PretendoNetwork/nex-protocols-go/blob/master/match-making/gathering_flags.go

jonbarrow commented on ranking/get_ranking.go:
```diff
@@ -92,3 +92,25 @@ func (protocol *Protocol) handleGetRanking(packet nex.PacketInterface) {
+
+// RankingModes is an enum of all the types of ranking request that can be made in the GetRanking method
+var RankingModes = rankingModes{
+	Global:        0,
+	NearbyGlobal:  1,
+	Friends:       2,
+	NearbyFriends: 3,
+	Self:          4,
```
For cases like this where we emulate an enum, we can avoid defining each value manually by using `iota` (which defaults to 0). See https://github.com/PretendoNetwork/nex-go/tree/master/constants for examples

ashquarky commented:
Rebased onto latest and added more of the enums used by Ranking - `RankingChangeAttributesParam.ModificationFlag`; `RankingMode` as before; `RankingScoreData.OrderBy`; `RankingOrderParam.OrderCalculation` and `RankingScoreData.UpdateMode`.

Based on the `match-making/constants` folder.

jonbarrow commented on ranking/constants/ranking_change_attribute_flag.go:
```diff
@@ -0,0 +1,11 @@
+// updated.
+type RankingChangeAttributeFlag uint8
+
+const (
+	RankingChangeAttributeFlagGroup1 RankingChangeAttributeFlag = 0x1
+	RankingChangeAttributeFlagGroup2                            = 0x2
+	RankingChangeAttributeFlagParam                             = 0x10
+)
```
```suggestion
const (
	// RankingChangeAttributeFlagNone indicates that no updates should occur
	RankingChangeAttributeFlagNone RankingChangeAttributeFlag = iota

	// RankingChangeAttributeFlagGroup0 indicates that the group at the 0th group should be updated
	RankingChangeAttributeFlagGroup0

	// RankingChangeAttributeFlagGroup1 indicates that the group at the 1st group should be updated
	RankingChangeAttributeFlagGroup1

	// RankingChangeAttributeFlagGroup2 indicates that the group at the 2nd group should be updated. Unused, only 2 groups are supported
	RankingChangeAttributeFlagGroup2

	// RankingChangeAttributeFlagGroup3 indicates that the group at the 3rd group should be updated. Unused, only 2 groups are supported
	RankingChangeAttributeFlagGroup3

	// RankingChangeAttributeFlagParam indicates that the param should be updated
	RankingChangeAttributeFlagParam
)
```

ashquarky replied:
:fire:

jonbarrow commented on ranking/constants/ranking_change_attribute_flag.go:
```diff
@@ -0,0 +1,11 @@
+package constants
+
+// RankingChangeAttributeFlag is used by RankingChangeAttributesParam.ModificationFlag to set which fields should be
+// updated.
+type RankingChangeAttributeFlag uint8
```
Shouldn't this just be named `ModificationFlag`?

ashquarky replied:
I'm not sure if we need to keep the `Ranking` prefix since it's in the global namespace?

jonbarrow replied:
I meant the part after the prefix. Though it being in this package is a good point. @DaniElectra what say you? If we remove the prefixes here we'd want to do it everywhere too

DaniElectra replied:
That sounds good to me. I don't think the prefix is used anywhere else though?

jonbarrow replied:
Just double checked, you're right I can't find anywhere that we use the prefix currently. So it should be dropped here

DaniElectra replied:
That shouldn't be an issue, since the import already tells that these are Ranking constants:

```go
import (
	ranking_constants "github.com/PretendoNetwork/nex-protocols-go/v2/ranking/constants"
)

ranking_constants.OrderByAscending
```

Also, this isn't a global namespace, the declarations are contained within the `ranking/constants` directory so there wouldn't be any collisions with external names

jonbarrow commented on ranking/constants/ranking_modes.go:
```diff
@@ -0,0 +1,21 @@
+// RankingMode represents the selection of who will be included on the leaderboard (global, friends, nearby etc.)
+type RankingMode uint32
+
+const (
+	// RankingModeGlobal is for everyone on the leaderboard, starting from global rank 1.
+	RankingModeGlobal RankingMode = iota
+
+	// RankingModeNearbyGlobal is the global leaderboard, but offset to center the selected user in the rankings.
+	RankingModeNearbyGlobal
+
+	// RankingModeFriends only has the users' friends on it (including themselves).
+	RankingModeFriends
+
+	// RankingModeNearbyFriends is the friends leaderboard, but offset to center the selected user in the rankings.
+	RankingModeNearbyFriends
+
+	// RankingModeSelf only returns the selected user's ranking.
+	RankingModeSelf
+)
```
```suggestion
const (
	// RankingModeRange is for everyone on the leaderboard, starting from global rank 1.
	RankingModeRange RankingMode = iota

	// RankingModeNear is the global leaderboard, but offset to center the selected user in the rankings.
	RankingModeNear

	// RankingModeFriendRange only has the users' friends on it (including themselves).
	RankingModeFriendRange

	// RankingModeFriendNEar is the friends leaderboard, but offset to center the selected user in the rankings.
	RankingModeFriendNEar

	// RankingModeUser only returns the selected user's ranking.
	RankingModeUser
)
```

jonbarrow commented on ranking/constants/ranking_order.go:
```diff
@@ -0,0 +1,9 @@
+package constants
+
+// RankingOrder is used in RankingScoreData.OrderBy to set the "golf scoring" mode for a category
+type RankingOrder uint8
```
Shouldnt this just be named `OrderBy`

jonbarrow commented on ranking/constants/ranking_update_mode.go:
```diff
@@ -0,0 +1,12 @@
+// RankingUpdateMode is used by RankingScoreData.UpdateMode to control if worse scores should be discarded or not.
+type RankingUpdateMode uint8
+
+const (
+	// RankingUpdateModeIfBetter will only submit the updated score if it's better than the old score in that category.
+	RankingUpdateModeIfBetter RankingUpdateMode = iota
+
+	// RankingUpdateModeAlways will always overwrite any old score, even if the new one is worse.
+	RankingUpdateModeAlways
+)
```
```suggestion
const (
	// RankingUpdateModeNormal will only commit the updated score if it's better than the old score in that category.
	RankingUpdateModeNormal RankingUpdateMode = iota

	// RankingUpdateModeDeleteOld will always overwrite any old score, even if the new one is worse.
	RankingUpdateModeDeleteOld
)
```

jonbarrow commented on ranking/constants/ranking_order_calculation.go:
```diff
@@ -0,0 +1,13 @@
+// RankingOrderCalculation is used in RankingOrderParam.OrderCalculation to control how ties are handled.
+type RankingOrderCalculation uint8
+
+const (
+	// RankingOrderCalculationStandard requests standard "1224" competition ranking.
+	RankingOrderCalculationStandard RankingOrderCalculation = iota
+
+	// RankingOrderCalculationOrdinal requests strictly ordinal "1234" ranking, with ties broken first by update time
+	// (earlier is better) then user PID (lower is better).
+	RankingOrderCalculationOrdinal
+)
```
```suggestion
const (
	// RankingOrderCalculation113 requests standard "1224" competition ranking.
	RankingOrderCalculation113 RankingOrderCalculation = iota

	// RankingOrderCalculation123 requests strictly ordinal "1234" ranking, with ties broken first by update time
	// (earlier is better) then user PID (lower is better).
	RankingOrderCalculation123
)
```

jonbarrow commented on ranking/constants/ranking_modes.go:
This file name is a plural, when the others are not?

jonbarrow reviewed (commented):
We know the real names of these values now, via https://github.com/tech-ticks/RTDXTools/blob/232a7797e01369e9c12704f58fdde65dd3ac1c32/Assets/Scripts/Stubs/Generated/Assembly-CSharp/NexPlugin/Ranking.cs

ashquarky commented:
Enums are ready for another look, happy to do a quick commit if we decide to drop the `Ranking` prefix

ashquarky commented:
In that case I think this is ready to merge
